[FEATURE] Script de création d'organisations PRO en masse depuis un fichier CSV (PIX-2278).

yahyayouness opened the pull request:
## :unicorn: Problème
L'équipe PRO doit créer les organisations PRO une par une. L'action devient très couteuse en temps lorsqu'il y en a plusieurs milliers.

## :robot: Solution
Ajouter un script qui permet de les créer en les important depuis un fichier csv en mode batch.

## :rainbow: Remarques
-Les données du fichier sont validées avant d'être insérées en base de donnée.
En cas d'erreur rencontrée, un rollback est effectuée afin ne pas corrompre les données (Utilisation du domainTransaction).
-La colonne tags contient des valeurs multiples séparées par un '_'.
-Pour optimiser le traitement, une recherche de tous les tags est effectuée une fois afin d'éviter les appels multiples de vérification unitaire de l'existence des tags.
-Deux appels en mode batch qui créent les organisations dans un premier temps, récupérés leurs identifiants afin de créer des organisations tags dans un second temps. Les deux appels séquentiels sont englobées dans une seule transaction.
❤️  mob team acces

## :100: Pour tester
Créer un fichier csv qui contient des lignes correspondantes aux informations de l'organisation.
Les colonnes correspondent respectivement au champs ci-dessous:
[ external Id, name, province Code, boolean canCollectProfiles, credit, email, locale, tags ]

Exemple:
b200;Entreprise pro;123;false;0;test@example.net;fr-fr;PUBLIC_AGRICULTURE

Vérifier la création de l'organisation ainsi que ses tags associés.

yahyayouness committed:
Add batchCreate organizations tags with transaction repository

pix-service commented:
I'm deploying this PR to these urls:

- App: https://app-pr2748.review.pix.fr
- Orga: https://orga-pr2748.review.pix.fr
- Certif: https://certif-pr2748.review.pix.fr
- Admin: https://admin-pr2748.review.pix.fr
- API: https://api-pr2748.review.pix.fr/api/

 Please check it out!

yahyayouness committed:
Add find all tags repository

MathieuGilet commented:
Il va manquer l'envoi d'invitations aux orgas dont on dispose d'une adresse e-mail. On ne l'avait pas implémenté en Mob mais c'est attendu dans le ticket.

chrc commented on api/lib/domain/usecases/create-pro-organizations.js:
```diff
@@ -0,0 +1,92 @@
+
+  await _checkIfOrganizationsAlreadyExistInDatabase(organizations, organizationRepository);
+
+  const csvData = _validateAndMapOrganizationsData(organizations);
```
On peut changer le nom de la variable

chrc commented on api/lib/domain/usecases/create-pro-organizations.js:
```diff
@@ -0,0 +1,92 @@
+const { ManyOrganizationsFoundError, OrganizationAlreadyExistError, OrganizationTagNotFound, ObjectValidationError } = require('../errors');
+const ORGANIZATION_TAG_SEPARATOR = '_';
+
+module.exports = async function createProOrganizations({
```
Pour le nom de ce usecase : on créé des organisations pro ET on leur associe des tags.

chrc commented on api/scripts/create-pro-organizations.js:
```diff
@@ -0,0 +1,82 @@
+
+function checkData({ csvData }) {
+  return csvData.map((data) => {
+    const [externalIdLowerCase, name, provinceCode, canCollectProfiles, credit, email, locale, tags ] = data;
```
Question : Pourquoi `externalIdLowerCase` ?

chrc commented on api/scripts/create-pro-organizations.js:
```diff
@@ -0,0 +1,82 @@
+    const [externalIdLowerCase, name, provinceCode, canCollectProfiles, credit, email, locale, tags ] = data;
+
+    if (!externalIdLowerCase && !name) {
+      if (require.main === module) process.stdout.write('Found empty line in input file.');
```
Le message d'erreur n'est pas en phase.

chrc commented on api/scripts/create-pro-organizations.js:
```diff
@@ -0,0 +1,82 @@
+    const filePath = process.argv[2];
+
+    console.log('Reading and parsing csv data file... ');
+    const csvData = await parseCsv(filePath);
```
Suggestion : L'utilisation du helper `parseCsvWithHeader` peut être intéressante (ce qui suppose d'utiliser un csv avec un header, et qui trim au passage les strings)

yahyayouness replied:
Done thanks

chrc commented on api/scripts/create-pro-organizations.js:
```diff
@@ -0,0 +1,82 @@
+
+    console.log('Checking data... ');
+    const organizations = checkData({ csvData });
+    console.log(organizations);
```
Question : Pourquoi afficher la liste des données ?

yahyayouness replied:
On nous a demandé de récupérer les ids des orgas créées, du coup on compte rediriger la sortie console vers un fichier pour les récupérer.

thloesch commented on api/scripts/create-pro-organizations.js:
```diff
@@ -0,0 +1,85 @@
+      return null;
+    }
+    if (!externalIdLowerCase) {
+      if (require.main === module) process.stdout.write(`A line is missing an externalId for name ${name}`);
```
**Suggestion** : 
```suggestion
      if (require.main === module) process.stdout.write(`\nA line is missing an externalId for name ${name}`);
```

Quand il y a plusieurs lignes problématiques, c'est un peu difficile à la lecture.
Je propose de rajouter un saut de ligne avant chaque message

thloesch commented on api/scripts/create-pro-organizations.js:
```diff
@@ -0,0 +1,85 @@
+const tagRepository = require('../lib/infrastructure/repositories/tag-repository');
+
+function checkData({ csvData }) {
+  return csvData.map((data) => {
```
**Remarque** : Avec ce map, on ne fait pas véritablement un check des données. On modifie les données pour enlever les lignes auxquelles il manque l'externalId ou le nom.

Ces données modifiées seront ensuite envoyées au usecase qui applique les même vérifications, sauf qu'on observe 2 comportements différents : 
- Si le script détecte une ligne problématique il la supprime, affiche un message en console, et continue son chemin
- Si le usecase en détecte, il interrompt immédiatement l'opération et renvoie une erreur.

Il me semble qu'on souhaitait bloquer si le fichier de base contenait ce genre de problème. 
Je trouve ça bien que les vérification se fassent côté script **et** côté usecase, mais le script devrait aussi interrompre l'opération à mon avis

yahyayouness replied:
Bien vu, les verifications ont été enlevées du script et centralisées dans le usecase.

thloesch commented on api/lib/infrastructure/repositories/organization-repository.js:
```diff
@@ -58,6 +60,11 @@ module.exports = {
       .then(_toDomain);
   },
 
+  async batchCreate(organizations, domainTransaction = DomainTransaction.emptyTransaction()) {
```
**Suggestion**

Comme discuté en mob, nous avons rajouté `credit` dans le pick, à la différence du `create` déjà existant.
Pour éviter les ambiguïté je propose :

```suggestion
  async batchCreateProOrganizations(organizations, domainTransaction = DomainTransaction.emptyTransaction()) {
```

thloesch commented:
**À discuter en récap tech**

Les règles de gestion concernant les valeurs par défaut ne sont pas testées explicitement dans le usecase

thloesch commented on api/lib/domain/usecases/create-pro-organizations.js:
```diff
@@ -0,0 +1,106 @@
+
+  const organizationIds = await organizationRepository.findByExternalIdsFetchingIdsOnly(map(organizations, 'externalId'));
+  if (!isEmpty(organizationIds)) {
+    throw new OrganizationAlreadyExistError();
```
**Remarque**

Cette erreur empêche la rejouabilité du script

thloesch replied:
Décidé en récap tech : On laisse comme ça pour cette PR sous condition que le checkData dans le script ne supprime plus de lignes, et on prévient @manux-pix

yahyayouness replied:
Le checkData a été enlevé pour une total vérification dans le usecase. Du coup on peut rejouer le script car à la moindre erreur, on rollback tout

yahyayouness committed:
Add batchCreate organizations with transaction repository

yahyayouness committed:
Create pro organizations invitations in service

yahyayouness committed:
Add OrganizationTagNotFound OrganizationAlreadyExist Errors

yahyayouness committed:
Add default value credit,canCollectProfiles for organization

+In order to test default values for (credit,canCollectProfiles) values in repository, we add this default values attributes into domain for the check
+Rename bachCreate to bachCreateOrganizationPro

yahyayouness committed:
Create pro organizations with tags and send invitations usecase

yahyayouness committed:
Create pro organizations with emails invitations in script

thloesch commented on api/lib/domain/models/Organization.js:
```diff
@@ -16,8 +21,8 @@ class Organization {
-    credit,
-    canCollectProfiles,
+    credit = defaultValues.credit,
+    canCollectProfiles = defaultValues.canCollectProfiles,
```
**Suggestion**

Cette règle métier n'est pas testée unitairement côté modèle.
Je propose de rajouter le test suivant :

```suggestion
    it('should build an Organization with default values for credit, canCollectProfiles when not specified', () => {
      // given
      const rawData = {
        id: 1,
        name: 'Lycée Jean Rostand',
        type: 'SCO',
      };

      // when
      const organization = new Organization(rawData);

      // then
      expect(organization.credit).to.equal(0);
      expect(organization.canCollectProfiles).to.equal(false);
    });
```

thloesch replied:
(Du coup ça ne veut pas dire qu'il faut enlever la ligne 25)

yahyayouness replied:
Yes, j'attendais juste la confirmation pour le rajouter :)

yahyayouness replied:
Done

thloesch commented on api/tests/integration/domain/usecases/create-pro-organization-with-tags_test.js:
```diff
@@ -0,0 +1,177 @@
+
+  });
+
+  it('should rollback create pro organizations with tags and throw an error when too many occurrences of the same organization', async () => {
```
**Non-bloquant**

```suggestion
  it('should rollback create pro organizations with tags and throw an error when there is more than one occurrences of the same organization', async () => {
```

thloesch commented on api/tests/integration/infrastructure/repositories/organization-repository_test.js:
```diff
@@ -823,4 +823,46 @@ describe('Integration | Repository | Organization', function() {
+      // then
+      const foundOrganizations = await knex('organizations').select();
+      expect(foundOrganizations.length).to.equal(1);
+      expect(foundOrganizations[0].canCollectProfiles).to.equal(Organization.defaultValues['canCollectProfiles']);
+      expect(foundOrganizations[0].credit).to.equal(Organization.defaultValues['credit']);
```
**Suggestion**

Merci d'avoir rajouté ce test !

Suite à nos discussions avec la team Prescription, ce test est utile pour vérifier les valeurs par défaut de la BDD lors d'une insertion d'organisation. 
Même si les valeurs sont les mêmes, je trouve que passer par les defautValues du modèle (donc côté domaine) pour faire cette vérification cache un peu cette intention.

Autre suggestion, je le déplacerais en dehors de notre test du `batchCreateProOrganizations`, car il concerne toutes les insertions

yahyayouness replied:
Déplacé thanks
